base_layer/p2p/src/saf_service/saf_service.rs

```diff
@@ -80,69 +115,168 @@ impl SAFService {
     }
 
     /// Send a message retrieval request to all neighbouring peers that are in the same network region.
-    fn send_retrieval_request(&self) -> Result<(), SAFError> {
-        let _oms = self.oms.as_ref().ok_or(SAFError::OMSUndefined)?;
-
-        // TODO: construct and send a message retrieval request to all neighbouring peers using oms
+    fn send_retrieval_request(&self, start_time: Option<DateTime<Utc>>) -> Result<(), SAFError> {
+        let oms = self.oms.as_ref().ok_or(SAFError::OMSUndefined)?;
+        let node_identity = self.node_identity.as_ref().ok_or(SAFError::NodeIdentityUndefined)?;
+
+        oms.send_message(
+            BroadcastStrategy::Closest(ClosestRequest {
+                n: DHT_BROADCAST_NODE_COUNT,
+                node_id: node_identity.identity.node_id.clone(),
+                excluded_peers: Vec::new(),
+            }),
+            MessageFlags::ENCRYPTED,
+            RetrieveMsgsMessage { start_time },
+        )?;
+        trace!(target: LOG_TARGET, "Message retrieval request sent");
 
         Ok(())
     }
 
     /// Process an incoming message retrieval request.
     fn receive_retrieval_request(&mut self, connector: &DomainConnector<'static>) -> Result<(), SAFError> {
-        let _oms = self.oms.as_ref().ok_or(SAFError::OMSUndefined)?;
+        let oms = self.oms.as_ref().ok_or(SAFError::OMSUndefined)?;
+        let peer_manager = self.peer_manager.as_ref().ok_or(SAFError::PeerManagerUndefined)?;
+        let node_identity = self.node_identity.as_ref().ok_or(SAFError::NodeIdentityUndefined)?;
 
-        let incoming_msg: Option<(MessageInfo, StoredMsgsMessage)> = connector
+        let incoming_msg: Option<(MessageInfo, RetrieveMsgsMessage)> = connector
             .receive_timeout(Duration::from_millis(1))
             .map_err(SAFError::ConnectorError)?;
-        if let Some((_info, _stored_msgs)) = incoming_msg {
+        if let Some((info, retrieval_request_msg)) = incoming_msg {
+            if peer_manager.in_network_region(
+                &info.peer_source.node_id,
+                &node_identity.identity.node_id,
+                DHT_BROADCAST_NODE_COUNT,
+            )? {
+                // Compile a set of stored messages for the requesting peer
+                // TODO: compiling the bundle of messages is slow, especially when there are many stored messages, a
+                // better approach should be used
+                let mut stored_msgs_response = StoredMsgsMessage {
+                    message_envelopes: Vec::new(),
+                };
+                for (_, stored_message) in self.msg_storage.iter_mut() {
```

Comment:
Is `iter_mut` needed - didn't spot any mutation

Reply:
Good catch, change applied